src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c

```diff
@@ -0,0 +1,381 @@
+	struct xrt_cu *xcu;
+
+	if (!cu)
+		return 0;
+
+	xcu = &cu->base;
+	return xrt_cu_circ_consume_all(xcu, buf, count);
+}
+
+static struct attribute *scu_attrs[] = {
+	&dev_attr_debug.attr,
+	&dev_attr_cu_stat.attr,
+	&dev_attr_cu_info.attr,
+	&dev_attr_poll_interval.attr,
+	&dev_attr_busy_threshold.attr,
+	&dev_attr_name.attr,
+	&dev_attr_base_paddr.attr,
+	&dev_attr_size.attr,
+	&dev_attr_stat.attr,
+	NULL,
+};
+
+static struct bin_attribute crc_buf_attr = {
+	.attr = {
+		.name = "crc_buf",
+		.mode = 0444
+	},
+	.read = crc_buf_show,
+	.write = NULL,
+	.size = 0,
+};
+
+static struct bin_attribute *scu_bin_attrs[] = {
+	&crc_buf_attr,
+	NULL,
+};
+
+static const struct attribute_group scu_attrgroup = {
+	.attrs = scu_attrs,
+	.bin_attrs = scu_bin_attrs,
+};
+
+static int scu_add_args(struct xocl_scu *xcu, struct kernel_info *kinfo)
+{
+	struct xrt_cu_arg *args = NULL;
+	int i;
+
+	/* If there is no detail kernel information, maybe it is a manualy
+	 * created xclbin. Print warning and let it go through
+	 */
+	if (!kinfo) {
+		XSCU_WARN(xcu, "CU %s metadata not found, xclbin maybe corrupted",
+			 xcu->base.info.iname);
+		xcu->base.info.num_args = 0;
+		xcu->base.info.args = NULL;
+		return 0;
+	}
+
+	args = vmalloc(sizeof(struct xrt_cu_arg) * kinfo->anums);
+	if (!args)
+		return -ENOMEM;
+
+	if (kinfo) {
+		for (i = 0; i < kinfo->anums; i++) {
+			strcpy(args[i].name, kinfo->args[i].name);
+			args[i].offset = kinfo->args[i].offset;
+			args[i].size = kinfo->args[i].size;
+			args[i].dir = kinfo->args[i].dir;
+		}
+		xcu->base.info.num_args = kinfo->anums;
+		xcu->base.info.args = args;
+	}
+
+	return 0;
+}
+
+static void scu_del_args(struct xocl_scu *xcu)
+{
+	if (xcu->base.info.args)
+		vfree(xcu->base.info.args);
+}
+
+static int scu_probe(struct platform_device *pdev)
+{
+	xdev_handle_t xdev = xocl_get_xdev(pdev);
+	struct xocl_scu *xcu;
+	struct resource **res;
+	struct xrt_cu_info *info;
+	struct kernel_info *krnl_info;
+	struct xrt_cu_arg *args = NULL;
+	int err = 0;
+	int i;
+
+	/* Not using xocl_drvinst_alloc here. Because it would quickly run out
+	 * of memory when there are a lot of cards. Since user cannot open CU
+	 * subdevice, the normal way to allocate device is good enough.
+	 */
+	xcu = devm_kzalloc(&pdev->dev, sizeof(*xcu), GFP_KERNEL);
+	if (!xcu)
+		return -ENOMEM;
+
+	xcu->pdev = pdev;
+	xcu->base.dev = XDEV2DEV(xdev);
+
+	info = XOCL_GET_SUBDEV_PRIV(&pdev->dev);
+	BUG_ON(!info);
+	memcpy(&xcu->base.info, info, sizeof(struct xrt_cu_info));
+
+	xcu->base.info.model = XCU_HLS;
+
+	krnl_info = xocl_query_kernel(xdev, info->kname);
+	err = scu_add_args(xcu, krnl_info);
+	if (err)
+		goto err;
+
+	res = vzalloc(sizeof(struct resource *) * xcu->base.info.num_res);
+	if (!res) {
+		err = -ENOMEM;
+		goto err;
+	}
+
+	for (i = 0; i < xcu->base.info.num_res; ++i) {
+		res[i] = platform_get_resource(pdev, IORESOURCE_MEM, i);
+		if (!res[i]) {
+			err = -EINVAL;
+			goto err1;
+		}
+	}
```

Comment:
A little bit confusing here. Why PS kernel has IORESOURCE_MEM? I need devices.h change to see if this make sense or not.